src/peer_select.cc

```diff
@@ -548,7 +549,7 @@ PeerSelector::noteIps(const Dns::CachedIps *ia, const Dns::LookupDetails &detail
         lastError = nullptr;
         if (fs->code == HIER_DIRECT) {
             lastError = new ErrorState(ERR_DNS_FAIL, Http::scServiceUnavailable, request, al);
-            lastError->dnsError = details.error;
+            lastError->dnsError = SBufToString(details.error);
```

Comment:
I am a bit concerned about the scope of this PR balooning out of control. I'll give it a try and push back if it does, leaving it instead to a followup PR

Reply:
> The idea to start working on this comes from ... an attempt to turn ipcache ... to a single std::unordered_map and std::nth_element for O(n) ageout instead

Hm... Perhaps I am missing something important, but that plan to sort an unordered map does not sound right. Should not we use ClpMap or a similar map+LRU_index combination instead?!

Reply:
> Hm... Perhaps I am missing something important, but a plan to sort[1](#user-content-fn-1-5dfd48221387c3baae028553105c21db) an unordered map does not sound right. Should not we use ClpMap or a similar map+LRU_index combination instead?!

We can, and could (I didn't remember about ClpMap when I started).
my experiment here is trading a lower constant O(1) at insertion time, which is needed to reorder the lru_list, in exchange of a different constant O(n) at ageout time - might be larger or smaller: my idea walks and copies once all elements in the cache, instead of only the oldest ones, but then it only has to free/fix one data structure, not two. Hard to say which path is more expensive although I'm happy to accept considering it a slight higher cost for the propsed solution.
Re: using ClpMap, it might end up being more expensive because it strictly enforces never going over quota, while current ipcache implementation will wait until lru eviction time and go over quota in the meanwhile

Reply:
Jut to avoid misunderstanding: This discussion is unrelated to this PR. I am happy to continue it here, but it does not affect this PR state AFAICT.

----

> my experiment here is trading a lower constant O(1) at insertion time, which is needed to reorder the lru_list, in exchange of a different constant O(n) at ageout time

O(1) is a indeed a constant with respect to n -- a fixed cost `c` that does not change with `n`.

O(n) is not a constant -- it is a cost or overhead that is grows linearly with `n` (i.e. a cost function in the shape of `a*n + b`).

It is true that in any specific case (i.e. some fixed `n`, `a`, `b`, and `c`), the actual values of those functions may have any relationship. However, all other factors being equal, we do _not_ want to trade a constant O(1) operations in ClpMap for O(n) operations in some new code (not to mention O(n * log(n)) costs from std::nth_element), just because that relationship might be in our favor in _some_ cases. Please do not go further down that route without securing support for such a violation of basic performance principles.

> ClpMap ... might end up being more expensive because it strictly enforces never going over quota, while current ipcache implementation will wait until lru eviction time and go over quota in the meanwhile

In Squid, spreading costs over all transactions is better than slowing down a few unlucky transactions to pay for a tiny speedup enjoyed by many lucky transactions. We cannot always spread the costs evenly, but we should favor algorithms that do.

Also, "going over quota" is a dangerous bug that can be abused by external traffic (and does increase support overheads). In most cases, the performance benefits of violating configuration are not worth the associated headaches.

Finally, another big ClpMap advantage is that it already exists in the official code. We spent a lot of time getting it there. By default, we should attempt to reuse it rather than work hard on avoiding it.

The existing ipcache code has problems. I do not know whether the current dlink_list-based code (i.e. lru_list and friends) can be easily replaced by ClpMap _before_ some of those problems are solved. However, I think we should plan for (and work towards) that eventual replacement unless there are some serious reasons to prefer a new/different cache organization. I do not see any such reasons in this exchange.